### PR TITLE
SimplifiedNullReturnFixer - handle nullable return types

### PR DESCRIPTION
**Pull request by Slamdunk**
Which is the right branch for this kind of improvement?

- [x] Don't fix nullable return types
- [x] Remove `risky` flag

### CODE REVIEW

**Comment by Slamdunk** [comment]
I would like to remove the `risky` flag, since the fixer wouldn't produce anymore erroneous code.

What do you think?

**Comment by Slamdunk** [comment]
@keradus ping, at least for labels 🙏

**Comment by keradus** [comment]
@Slamdunk , sorry for big delay, simply too many things recently and i hoped someone else will pick the thread up (I was active in threads I participated previously, but was trying to not dive in too much threads at same time :( )

**Comment by keradus** [comment]
label/milestone:
it's debatable and I could advocate both approaches, but so far we treated "better support of newer language syntax" as bugfixes and sent those to lowest supported branch (currently: 2.2)

File: tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php [hunk 1]
```diff
@@ -55,4 +55,44 @@ public function provideFixCases()
...
+            [
+                '<?php function foo(): ?int { $bar = function() { return; }; return null; }',
+                '<?php function foo(): ?int { $bar = function() { return null; }; return null; }',
+            ],
+            [
+                '<?php function foo(): int { return; }',
+                '<?php function foo(): int { return null; }',
+            ],
+            [
+                '<?php function foo(): void { return; }',
```
Comment (keradus) [inline tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php]: please add a case with pure `: null` (not `:?...`) as well

File: tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php [hunk 2]
```diff
@@ -55,4 +55,44 @@ public function provideFixCases()
...
+            [
+                '<?php function foo() { return; } function bar(): ?A\B\C\D { return null; } function baz() { return; }',
+                '<?php function foo() { return null; } function bar(): ?A\B\C\D { return null; } function baz() { return null; }',
+            ],
+            [
+                '<?php function foo(): ?int { $bar = function() { return; }; return null; }',
+                '<?php function foo(): ?int { $bar = function() { return null; }; return null; }',
+            ],
+            [
+                '<?php function foo(): int { return; }',
```
Comment (keradus) [inline tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php]: > Fatal error: A function with return type must return a value

File: src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php [hunk 3]
```diff
@@ -101,21 +102,48 @@ private function clear(Tokens $tokens, $index)
...
+        $content = rtrim($content, ');');
 
         return 'null' === strtolower($content);
     }
 
     /**
+     * Is the return within a function with a nullable return type?
+     *
+     * @param Tokens $tokens
+     * @param int    $index
```
Comment (keradus) [inline src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: this is an index of ?

File: tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php (hunk 2 above)
Comment (Slamdunk) [reply tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php]:
```
function foo(): int { return null; }
// Fatal error: Uncaught TypeError: Return value of foo() must be of the type integer, null returned

function foo(): int { return; }
// Fatal error: A function with return type must return a value
```
Both before and after the fix there is a Fatal error.

The reason why I wrote this test is to prove **this** PR doesn't change the (erroneous) behaviour of the fixer.

File: src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php (hunk 3 above)
Comment (Slamdunk) [reply src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: Just copied from the other fixer methods. Btw I've added the  description now.

**Comment by Slamdunk** [comment]
> label/milestone:
> it's debatable and I could advocate both approaches, but so far we treated "better support of newer language syntax" as bugfixes and sent those to lowest supported branch (currently: 2.2)
> (in the end, currently 2.2 removes the retrurn **null** where it shouldn't)
> 
> WDYT?

My request for label/milestone was just to set some reminders. I respect your great efforts and time spent in this project, I don't expect a rapid merge, just that without labels it's easy to lose the plot.

Rebased onto 2.2, I agree.

File: tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php (hunk 2 above)
Comment (keradus) [reply tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php]: I agree that initial code was wrong as well. IMO we shall not change one wrong code into another, but simply don't touch it.

Comment (Slamdunk) [reply tests/Fixer/ReturnNotation/SimplifiedNullReturnFixerTest.php]: Done.

File: src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php [hunk 4]
```diff
@@ -101,20 +102,51 @@ private function clear(Tokens $tokens, $index)
...
+    {
+        $functionIndex = $returnIndex;
+        do {
+            $functionIndex = $tokens->getPrevTokenOfKind($functionIndex, array(array(T_FUNCTION)));
+            if (null === $functionIndex) {
+                return false;
+            }
+            $openingCurlyBraceIndex = $tokens->getNextTokenOfKind($functionIndex, array('{'));
+            $closingCurlyBraceIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $openingCurlyBraceIndex);
+        } while ($closingCurlyBraceIndex < $returnIndex);
```
Comment (keradus) [inline src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: ha, so that's how you skip lambdas ?

Comment (keradus) [reply src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: finding `closingCurlyBraceIndex` is expensive, we don't care about that information for lambdas, you use it here only to determine if a `T_FUNCTION` a lambda or not.
what about using `TokensAnalyzer::isLambda` instead?

**Comment by keradus** [comment]
for me, now the fixer doesn't look risky indeed.

File: src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php (hunk 4 above)
Comment (Slamdunk) [reply src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: The purpose here is different.
Keep in mind that this fixer starts its algorithm from return keyword: what method signature this return belongs to?
Now imagine:

* a class method with a nullable return type
* that contains an anonymous class with a non-nullable method
* that contains an anonymous class with a nullable method
* that contains an anonymous class with a non-nullable method

If in this Matryoshka doll each method also has a return statement, we need this loop to determine each correspondent method signature.

The second-last test case represent this situation.

Comment (keradus) [reply src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: whole method, yes.
But, isn't the `do... while` construction just to skip the content of internal lambda ?

Comment (Slamdunk) [reply src/Fixer/ReturnNotation/SimplifiedNullReturnFixer.php]: No, it's to find out the function's signature the return belongs to.
It goes backward, if the function ends before return skip it (aka keep going backward), if the function ends after return use it.
It doesn't matter which type of function it is.

**Comment by Slamdunk** [comment]
Riskyness removed

**Comment by Slamdunk** [comment]
Since the removal of riskyness, it wasn't clear enough what now this Fixer does. I've added a `VersionSpecificCodeSample` to be clear.

**Comment by keradus** [comment]
Thank you @Slamdunk.

**Pull request closed without merging**